gnupatch: 2.7.6 -> 2.7.6-unstable-2024-08-25

tie opened the pull request:
## Description of changes

This change updates `gnupatch` to the latest Git commit.

Note that stdenv/linux uses fetchurlBoot as fetchurl for bootstrap and we need fetchFromSavannah (or rather fetchzip) that is not compatible with fetchurlBoot, so we override fetchzip to use fetchurl from the current package set instead of using fetchurlBoot. This shouldn’t be an issue on Darwin because curl is included in bootstrap files. FreeBSD stdenv seems to have curl in bootstrap files as well.

**References:**
- https://savannah.gnu.org/bugs/?57883#comment2
- https://github.com/NixOS/nixpkgs/pull/337473
- https://github.com/NixOS/nixpkgs/pull/335579#discussion_r1722013481

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

tie committed:
gnulibBootstrapHook: init

tie committed:
gnulib: 20231109 -> 1.0

tie committed:
gnupatch: 2.7.6 -> 2.7.6-unstable-2024-08-25

emilazy commented:
Thanks! I don’t feel entirely confident to review this gnarly stdenv bootstrap stuff but I’ll try to take a closer look soon.

Could we avoid some of the bootstrapping issues by just using `fetchurl`? I don’t know if we need `fetchzip` specifically for this.

tie commented:
>Could we avoid some of the bootstrapping issues by just using fetchurl? I don’t know if we need fetchzip specifically for this.

If Git tarballs generated by Cgit are reproducible, then yes. Otherwise we can’t use `fetchurlBoot` because it hashes the file instead of archive contents. E.g. see https://github.blog/open-source/git/update-on-the-future-stability-of-source-code-archives-and-hashes

tie committed:
pkgs/stdenv/linux: bootstrap fetchurl earlier

This change bootstraps fetchurl earlier in the pkgs/stdenv/linux stages
for use with other fetchers like fetchzip.

tie committed:
openssl: use makeBinaryWrapper instead of makeShellWrapper

This changes openssl to use makeBinaryWrapper since makeWrapper uses
non-overridable runtimeShell that causes infinite recursion. That is,
fetchurl in pkgs/top-level/all-packages.nix is bootstrapped by
overriding dependencies to use stdenv.fetchurlBoot.

This change potentially allows us to drop curl from bootstrap files
for other platforms.

philiptaron commented on pkgs/top-level/all-packages.nix:
```diff
@@ -184,6 +184,25 @@ with pkgs;
     autoconf = autoconf271;
   };
 
+  gnulibBootstrapHook = callPackage (
+    {
+      makeSetupHook,
+      gnulib,
+      autoconf,
+      automake,
+    }:
+    makeSetupHook {
+      name = "gnulib-bootstrap-hook";
+      propagatedBuildInputs = [
+        autoconf
+        automake
+      ];
+      substitutions = {
+        inherit gnulib;
+      };
+    } ../build-support/setup-hooks/gnulib-bootstrap.sh
+  ) { };
```
I _think_ this can be in `pkgs/by-name/gn/gnulibBootstrapHook`. I'd prefer that.

tie replied:
Hm, I’d at least expect pkgs/by-name to be case-insensitive (or forcing all to be lower-case) since that’s a reasonable behavior given that macOS and Windows use case-insensitive filesystems by default. I’m not sure if that’s the case though.

philiptaron replied:
We have a check in `nixpkgs-vet` that names can't differ by case -- `food` and `FooD` are rejected. macOS and Windows are case insensitive but case-_preserving_, so we can use the filesystem for determining attrset names as `pkgs/by-name` does.

philiptaron commented on pkgs/stdenv/linux/default.nix:
```diff
@@ -624,7 +633,7 @@ in
       shell = cc.shell;
 
-      inherit (prevStage.stdenv) fetchurlBoot;
+      inherit fetchurlBoot;
```
Why did this need to change?

tie replied:
No reason in particular, it was easier to remove all fetchurl/fetchurlBoot instances in stdenv/linux and rewrite it from scratch to when working on this PR, and I haven’t minimized the diff yet because some commits should be split into separate PRs (e.g. openssl stuff).

philiptaron commented on pkgs/development/libraries/openssl/default.nix:
I think we can merge this separate from this PR.

tie replied:
Yes, I was planning to do this 👍

philiptaron commented on pkgs/development/tools/gnulib/default.nix:
I think we can merge this separate from this PR, and since it's a version bump, I'd prefer that.

emilazy replied:
Can we? I thought the change to `python3Minimal` would be load‐bearing.

tie replied:
I’m not really sure if gnulib even uses/needs python3, i.e. by default gnulib-tool does a PATH lookup before using Python implementation (unless `GNULIB_TOOL_IMPL` is explicitly set).

https://github.com/coreutils/gnulib/blob/6e95321249ae6986c3df764a6e539f8b5be13948/gnulib-tool#L161

philiptaron commented:
https://github.com/NixOS/nixpkgs/actions/runs/10648732824/job/29518094549?pr=337961

This `nixpkgs-vet` error is unexpected to me. Taking a look.

emilazy commented:
I think the check was broken at some point? It might need a rebase.

infinisil commented:
The error is misleading, but the underlying problem seems to be that `pkgs.gnupatch` isn't actually pointing to the definition in `pkgs/by-name`, neither before nor after this PR:

```
nix-repl> builtins.unsafeGetAttrPos "gnupatch" pkgs
{
  column = 33;
  file = "/home/ncasil/src/nixpkgs/pkgs/stdenv/linux/default.nix";
  line = 679;
}
```

Not entirely sure why that happens, because both `all-packages.nix` and `pkgs/by-name` should override stdenv definitions: https://github.com/NixOS/nixpkgs/blob/3c799ef618f8a0ea383898b7e8877fcec4783988/pkgs/top-level/stage.nix#L348-L355

tie commented:
@infinisil, I assume this is because of the `stdenv.overrides` inherited from the previous stage (and `stdenvOverrides` taking precedence over `allPackages`)?

<details>

https://github.com/NixOS/nixpkgs/blob/c2090cb3038be415e3977efaaa60d5e8b4411402/pkgs/stdenv/linux/default.nix#L667-L701
https://github.com/NixOS/nixpkgs/blob/c2090cb3038be415e3977efaaa60d5e8b4411402/pkgs/top-level/stage.nix#L161-L165
https://github.com/NixOS/nixpkgs/blob/c2090cb3038be415e3977efaaa60d5e8b4411402/pkgs/top-level/stage.nix#L352-L363

</details>

This feels really weird to me, especially since the inherited (g)libc depends on `xgcc-*-libgcc` while other packages (that are not present in `stdenv.overrides`) are using regular `gcc-*-libgcc`. E.g.
```console
$ nix path-info --recursive $(nix build --no-link --print-out-paths --file . bash^out) | fgrep libgcc
/nix/store/3kghvsns26scklmgi2f14787lzwrvqnf-xgcc-13.3.0-libgcc
```
```console
nix path-info --recursive $(nix build --no-link --print-out-paths --file . coreutils^out) | fgrep libgcc
/nix/store/3kghvsns26scklmgi2f14787lzwrvqnf-xgcc-13.3.0-libgcc
/nix/store/vxk9halw286r11jabdyni93kkghhb9f5-gcc-13.3.0-libgcc
```
But perhaps that is a separate issue.

wegank commented:
gnupatch has been updated to 2.8 in #398071. If there's a need for bumping to an unstable version, please open a new PR.
